### ISSUE TITLE
Docs issue report from "features/html/general-html-support.html"

### ISSUE THREAD

**Issue by a-lam** [ckeditor/ckeditor5#15355]
### Origin URL

https://ckeditor.com/docs/ckeditor5/latest/features/html/general-html-support.html

### Project version

40.1.0

### Is there something missing in the guide? What is it?

With an Angular implementation using TypeScript, following the documented examples for configuring htmlSupport.allow causes compilation errors when you try to use webpack (or just `npm run build`) to packakge CKEditor after configuring.

Configuration attempted (as provided by the documentation):
```
htmlSupport: {
    allow: [
        {
            name: /.*/,
            attributes: true,
            classes: true,
            styles: true
        }
    ]
}
```

I receive the following error.
```
      TS2417: Class static side 'typeof Editor' incorrectly extends base class static side 'typeof ClassicEditor'.
  The types of 'defaultConfig.htmlSupport.allow' are incompatible between these types.
    Type '{ name: RegExp; attributes: boolean; classes: boolean; styles: boolean; }[]' is not assignable to type 'MatcherObjectPattern[]'.
      Type '{ name: RegExp; attributes: boolean; classes: boolean; styles: boolean; }' is not assignable to type 'MatcherObjectPattern'.
        Types of property 'classes' are incompatible.
          Type 'boolean' is not assignable to type 'ClassPatterns | undefined'.
ts-loader-default_e3b0c44298fc1c14

webpack 5.89.0 compiled with 1 error in 13503 ms
```

It is required to use `true as const` in the configuration, which allows the project to compile.
```
    htmlSupport: {
      allow: [
        {
          name: /.*/,
          attributes: true as const,
          classes: true as const,
          styles: true as const
        }
      ]
    }
```

It would save individuals a lot of confusion if the documentation was updated to add a note or mention that a user must use `true as const` if configuring a TypeScript project.

### User agent

Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36

**Comment by Witoso** [comment]
Thanks for reaching out. Please share how do you pass the config to the editor. Is it set in the build, or do you set it in your Angular component?

**Comment by a-lam** [comment]
In the build downloaded from the CKEditor site, I was modifying the defaultConfig in the src/ckeditor.ts. 
```typescript
/**
 * @license Copyright (c) 2014-2023, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
 */

import { ClassicEditor } from '@ckeditor/ckeditor5-editor-classic';

import { UploadAdapter } from '@ckeditor/ckeditor5-adapter-ckfinder';
import { Alignment } from '@ckeditor/ckeditor5-alignment';
import { Bold, Italic, Strikethrough, Underline } from '@ckeditor/ckeditor5-basic-styles';
import { BlockQuote } from '@ckeditor/ckeditor5-block-quote';
import { CKFinder } from '@ckeditor/ckeditor5-ckfinder';
import { Essentials } from '@ckeditor/ckeditor5-essentials';
import { FindAndReplace } from '@ckeditor/ckeditor5-find-and-replace';
import { FontBackgroundColor, FontColor, FontFamily, FontSize } from '@ckeditor/ckeditor5-font';
import { Heading } from '@ckeditor/ckeditor5-heading';
import { HorizontalLine } from '@ckeditor/ckeditor5-horizontal-line';
import { GeneralHtmlSupport } from '@ckeditor/ckeditor5-html-support';
import {
    Image,
    ImageResize,
    ImageStyle,
    ImageToolbar,
    ImageUpload
} from '@ckeditor/ckeditor5-image';
import { Indent } from '@ckeditor/ckeditor5-indent';
import { Link } from '@ckeditor/ckeditor5-link';
import { List } from '@ckeditor/ckeditor5-list';
import { Paragraph } from '@ckeditor/ckeditor5-paragraph';
import { PasteFromOffice } from '@ckeditor/ckeditor5-paste-from-office';
import { RemoveFormat } from '@ckeditor/ckeditor5-remove-format';
import { SelectAll } from '@ckeditor/ckeditor5-select-all';
import { SourceEditing } from '@ckeditor/ckeditor5-source-editing';
import {
    SpecialCharacters,
    SpecialCharactersEssentials,
    SpecialCharactersText
} from '@ckeditor/ckeditor5-special-characters';
import { Table, TableProperties, TableToolbar } from '@ckeditor/ckeditor5-table';
import { TextTransformation } from '@ckeditor/ckeditor5-typing';
import { EditorWatchdog } from '@ckeditor/ckeditor5-watchdog';

// You can read more about extending the build with additional plugins in the "Installing plugins" guide.
// See https://ckeditor.com/docs/ckeditor5/latest/installation/plugins/installing-plugins.html for details.

class Editor extends ClassicEditor {
  public static override builtinPlugins = [
    Alignment,
    BlockQuote,
    Bold,
    CKFinder,
    Essentials,
    FindAndReplace,
    FontBackgroundColor,
    FontColor,
    FontFamily,
    FontSize,
    GeneralHtmlSupport,
    Heading,
    HorizontalLine,
    Image,
    ImageResize,
    ImageStyle,
    ImageToolbar,
    ImageUpload,
    Indent,
    Italic,
    Link,
    List,
    Paragraph,
    PasteFromOffice,
    RemoveFormat,
    SelectAll,
    SourceEditing,
    SpecialCharacters,
    SpecialCharactersEssentials,
    SpecialCharactersText,
    Strikethrough,
    Table,
    TableProperties,
    TableToolbar,
    TextTransformation,
    Underline,
    UploadAdapter
  ];

  public static override defaultConfig = {
    licenseKey: '',
    toolbar: {
      items: [
        'heading',
        'fontFamily',
        'fontSize',
        'fontColor',
        'fontBackgroundColor',
        '|',
        'bold',
        'italic',
        'strikethrough',
        'link',
        'alignment',
        'bulletedList',
        'numberedList',
        'specialCharacters',
        'removeFormat',
        '|',
        'selectAll',
        'sourceEditing',
        'outdent',
        'indent',
        'findAndReplace',
        '|',
        'horizontalLine',
        'CKFinder',
        'blockQuote',
        'insertTable',
        'undo',
        'redo'
      ]
    },
    language: 'en',
    image: {
      toolbar: [
        'imageTextAlternative',
        'imageStyle:inline',
        'imageStyle:block',
        'imageStyle:side'
      ]
    },
    table: {
      contentToolbar: [
        'tableColumn',
        'tableRow',
        'mergeTableCells',
        'tableProperties'
      ]
    },
    htmlSupport: {
      allow: [
        {
          name: /.*/,
          attributes: true as const,
          classes: true as const,
          styles: true as const
        }
      ]
    }
  };
}

export default { Editor, EditorWatchdog };

```

In my component, we reference the Editor class.
```html
<ckeditor [editor]="Editor"
          [(ngModel)]="model"
          *ngIf="!readonly"
          (change)="onChange($event)">
</ckeditor>
```

```typescript
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { WindowRef } from '../services';
import { ChangeEvent } from '@ckeditor/ckeditor5-angular/ckeditor.component';
import Editor from 'ckeditor5-custom-build/build/ckeditor';

@Component({
  selector: 'my-editor',
  templateUrl: './my-editor.component.html'
})
export class MyEditor {
  public Editor = Editor.Editor;
  @Input() model: any;
  @Input() readonly: boolean = false;

  @Output() onDataChange = new EventEmitter<any>();

  winRef: WindowRef;
  retrievedData: string = null;

  constructor(winRef: WindowRef) {
    this.winRef = winRef;
  }

  public onChange({ editor }: ChangeEvent) {
    if (editor) {
      const data = editor.data.get();
      this.onDataChange.emit(data);
    }
  }
}
```

**Comment by Witoso** [comment]
I wonder why this `as const` needs to be used. Calling in @filipsobol.

**Comment by filipsobol** [comment]
### TL;DR version

You must specify the type of the `defaultConfig' object like this:

```ts
import type { EditorConfig } from '@ckeditor/ckeditor5-core';

export class Editor extends ClassicEditor {
 //                                    VVVVVVVVVVVV
 public static override defaultConfig: EditorConfig = {};
}
```

### Explanation

Let's take this code as an example:

```js
import { ClassicEditor } from '@ckeditor/ckeditor5-editor-classic';
import { GeneralHtmlSupport } from '@ckeditor/ckeditor5-html-support';

export class Editor extends ClassicEditor {
 public static override builtinPlugins = [
  GeneralHtmlSupport
 ];

 public static override defaultConfig = {
  htmlSupport: {
   allow: [
    {
     name: /.*/,
     attributes: true,
     classes: true,
     styles: true
    }
   ]
  }
 };
}
```

Since the `defaultConfig` is not typed in this file, the TypeScript will [infer its type based on some rules](https://www.typescriptlang.org/docs/handbook/type-inference.html). It will make some guesses that will result in the following type:

```text
{
  htmlSupport: {
    allow: {
      name: RegExp;
      attributes: boolean;
      classes: boolean;
      styles: boolean;
    }[];
  }
}
```

As you can see, `true` was inferred as `boolean`, which is a good guess in most cases. Unfortunately, we don't accept `boolean` but `true` (or string, RegExp, etc.), which results in a TypeScript error.

This can be fixed by specifying the type of the `defaultConfig` object:

```ts
import type { EditorConfig } from '@ckeditor/ckeditor5-core';

export class Editor extends ClassicEditor {
  //                                    VVVVVVVVVVVV
  public static override defaultConfig: EditorConfig = {};
}
```

or

```ts
import type { EditorConfig } from '@ckeditor/ckeditor5-core';

export class Editor extends ClassicEditor {
  //                                        VVVVVVVVVVVVVVVVVVVVVV
  public static override defaultConfig = {} satisfies EditorConfig;
}
```

**Comment by filipsobol** [comment]
We should update our documentation if we don't have this information already. Maybe we can consider showing JavaScript and TypeScript code samples in the new installation docs? @godai78

**Comment by Witoso** [comment]
This is output from online builder, this is most likely the first place we should update :grimacing: @pomek

**Comment by godai78** [comment]
I believe coming up with code snippets that showcase TypeScript (by default) and JavaScript (alternate) as tabbed solution would be highly beneficial, but would also require some logic to employ. Considering this for new docs would be a nice challenge.

**Comment by Witoso** [comment]
Angular is very most popular in our integrations, most likely people went through `@ts-expect-error` route somewhere or GHS was not present in their setup.

**Comment by pomek** [comment]
Builds produced by Online Builder mark the `defaultConfig` property typed as `EditorConfig`.

![image](https://github.com/ckeditor/ckeditor5/assets/2270764/3b5816f9-3920-476c-86ab-cccbdf605416)

@Witoso, could we close the issue?
